### PR TITLE
Entrance Macro

### PR DESCRIPTION
**Pull request by hensldm**
Introduces the `ENTRANCE` macro which can be used to bitpack all the information stored in Entrances. Also does some small cleanup around those areas.

### CODE REVIEW

File: src/overlays/actors/ovl_En_Baba/z_en_baba.c [hunk 1]
```diff
@@ -510,7 +510,7 @@ void EnBaba_FinishInit(EnBaba* this, PlayState* play) {
         this->actionFunc = EnBaba_Idle;
     } else if (play->sceneNum == SCENE_BACKTOWN) {
         if ((BOMB_SHOP_LADY_GET_TYPE(&this->actor) == BOMB_SHOP_LADY_TYPE_FOLLOW_SCHEDULE) &&
-            (gSaveContext.save.entranceIndex != 0xD670) && (BOMB_SHOP_LADY_GET_PATH_INDEX(&this->actor) != 0x3F)) {
+            (gSaveContext.save.entrance != 0xD670) && (BOMB_SHOP_LADY_GET_PATH_INDEX(&this->actor) != 0x3F)) {
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Baba/z_en_baba.c:513]:
```suggestion
            (gSaveContext.save.entrance != ENTRANCE(NORTH_CLOCK_TOWN, 7, 0)) && (BOMB_SHOP_LADY_GET_PATH_INDEX(&this->actor) != 0x3F)) {
```

File: src/overlays/actors/ovl_En_Invadepoh/z_en_invadepoh.c [hunk 2]
```diff
@@ -1758,7 +1758,7 @@ void EnInvadepoh_InitRomani(EnInvadepoh* this, PlayState* play) {
         }
     } else if (temp != 8) {
         if (temp == 9) {
-            if (gSaveContext.save.entranceIndex != 0x6460) {
+            if (gSaveContext.save.entrance != 0x6460) {
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Invadepoh/z_en_invadepoh.c:1761]:
```suggestion
            if (gSaveContext.save.entrance != ENTRANCE(ROMANI_RANCH, 6, 0)) {
```

File: src/overlays/actors/ovl_En_Railgibud/z_en_railgibud.c [hunk 3]
```diff
@@ -1113,7 +1113,7 @@ void EnRailgibud_InitCutsceneGibdo(EnRailgibud* this, PlayState* play) {
     Collider_SetCylinder(play, &this->collider, &this->actor, &sCylinderInit);
     CollisionCheck_SetInfo2(&this->actor.colChkInfo, &sDamageTable, &sColChkInfoInit);
 
-    if (gSaveContext.save.entranceIndex != 0x2090) { // NOT Cutscene: Music Box House Opens
+    if (gSaveContext.save.entrance != 0x2090) { // NOT Cutscene: Music Box House Opens
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Railgibud/z_en_railgibud.c:1116]:
```suggestion
    if (gSaveContext.save.entrance != ENTRANCE(IKANA_CANYON, 9, 0)) { // NOT Cutscene: Music Box House Opens
```

File: src/overlays/actors/ovl_En_Pamera/z_en_pamera.c [hunk 4]
```diff
@@ -155,12 +155,13 @@ void EnPamera_Init(Actor* thisx, PlayState* play) {
...
-            (gSaveContext.save.weekEventReg[75] & 0x20) || (gSaveContext.save.entranceIndex == 0x2090)) {
+            (gSaveContext.save.weekEventReg[75] & 0x20) ||
+            (gSaveContext.save.entrance == ENTRANCE(IKANA_CANYON, 9, 0))) {
             Actor_MarkForDeath(&this->actor);
         }
         if (gSaveContext.save.weekEventReg[61] & 4) {
-            if (!(gSaveContext.save.weekEventReg[59] & 1) || (gSaveContext.save.entranceIndex != 0x2020)) {
-                if ((gSaveContext.save.entranceIndex != 0x2020) && (gSaveContext.save.weekEventReg[59] & 1)) {
+            if (!(gSaveContext.save.weekEventReg[59] & 1) || (gSaveContext.save.entrance != 0x2020)) {
+                if ((gSaveContext.save.entrance != 0x2020) && (gSaveContext.save.weekEventReg[59] & 1)) {
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Pamera/z_en_pamera.c:164]:
```suggestion
            if (!(gSaveContext.save.weekEventReg[59] & 1) || (gSaveContext.save.entrance != ENTRANCE(IKANA_CANYON, 2, 0))) {
                if ((gSaveContext.save.entrance != ENTRANCE(IKANA_CANYON, 2, 0)) && (gSaveContext.save.weekEventReg[59] & 1)) {
```

File: src/overlays/actors/ovl_En_Shn/z_en_shn.c [hunk 5]
```diff
@@ -356,7 +356,7 @@ void EnShn_Init(Actor* thisx, PlayState* play) {
     this->unk_2E0 = 0;
     this->unk_2D8 = 0;
     this->unk_1D8 = 0;
-    if (gSaveContext.save.entranceIndex != 0xA820) {
+    if (gSaveContext.save.entrance != 0xA820) {
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Shn/z_en_shn.c:359]: Missed this one

File: src/code/z_scene.c [hunk 6]
```diff
@@ -589,13 +589,13 @@ s32 Scene_ProcessHeader(PlayState* play, SceneCmd* header) {
 /**
  * Creates an entrance index from the scene index, spawn index, and scene setup.
  */
-u16 Entrance_CreateIndex(s32 sceneIndex, s32 spawnIndex, s32 sceneSetup) {
-    return (((sceneIndex << 9) | (spawnIndex << 4)) | sceneSetup) & 0xFFFF;
+u16 Entrance_CreateIndex(s32 scene, s32 spawn, s32 layer) {
+    return (scene << 9) | (spawn << 4) | layer;
 }
```
Comment (engineer124) [inline src/code/z_scene.c:594]:
```suggestion
u16 Entrance_CreateEntry(s32 scene, s32 spawn, s32 layer) {
    return (scene << 9) | (spawn << 4) | layer;
}
```
I wonder if this should become something like this, now that it's not `entranceIndex`. Same thing for `Entrance_CreateIndexFromSpawn`

File: include/z64scene.h [hunk 7]
```diff
@@ -650,6 +650,127 @@ typedef enum {
...
+    /* 0x66 */ ENTR_SCENE_GIANTS_CHAMBER,
+    /* 0x67 */ ENTR_SCENE_GORMAN_TRACK,
+    /* 0x68 */ ENTR_SCENE_GORON_RACETRACK,
+    /* 0x69 */ ENTR_SCENE_EAST_CLOCK_TOWN,
+    /* 0x6A */ ENTR_SCENE_WEST_CLOCK_TOWN,
+    /* 0x6B */ ENTR_SCENE_NORTH_CLOCK_TOWN,
+    /* 0x6C */ ENTR_SCENE_SOUTH_CLOCK_TOWN,
+    /* 0x6D */ ENTR_SCENE_LAUNDRY_POOL,
+    /* 0x6E */ ENTR_SCENE_MAX
+} EntranceSceneID;
```
Comment (engineer124) [inline include/z64scene.h:765]:
```suggestion
} EntranceSceneId;
```

**Review by jpburnett — COMMENTED** [review]
Seems like jenkins is complaining about new warnings: `Total new warnings: 3`

File: src/code/z_scene.c (hunk 6 above)
Comment (hensldm) [reply src/code/z_scene.c:594]: Yeah, that makes sense.

File: include/z64save.h [hunk 8]
```diff
@@ -121,7 +121,7 @@ typedef struct SavePlayerData {
 } SavePlayerData; // size = 0x28
 
 typedef struct Save {
-    /* 0x0000 */ u32 entranceIndex;                     // "scene_no"
+    /* 0x0000 */ u32 entrance;                     // "scene_no"
     /* 0x0004 */ u8 equippedMask;                       // "player_mask"
```
Comment (engineer124) [inline include/z64save.h:125]: Could you fix the spacing here for `// "scene_no"`?

File: src/code/z_scene.c [hunk 9]
```diff
@@ -589,13 +589,13 @@ s32 Scene_ProcessHeader(PlayState* play, SceneCmd* header) {
 /**
  * Creates an entrance index from the scene index, spawn index, and scene setup.
```
Comment (engineer124) [inline src/code/z_scene.c:590]:
```suggestion
 * Creates an entrance from the scene index, spawn index, and scene setup.
```

File: src/code/z_scene.c [hunk 10]
```diff
@@ -589,13 +589,13 @@ s32 Scene_ProcessHeader(PlayState* play, SceneCmd* header) {
...
  */
-u16 Entrance_CreateIndex(s32 sceneIndex, s32 spawnIndex, s32 sceneSetup) {
-    return (((sceneIndex << 9) | (spawnIndex << 4)) | sceneSetup) & 0xFFFF;
+u16 Entrance_CreateEntry(s32 scene, s32 spawn, s32 layer) {
+    return (scene << 9) | (spawn << 4) | layer;
 }
 
 /**
  * Creates an entrance index from the current entrance index with the given spawn index.
```
Comment (engineer124) [inline src/code/z_scene.c:597]:
```suggestion
 * Creates an entrance from the current entrance with the given spawn index.
```

File: src/code/z_scene.c (hunk 9 above)
Comment (hensldm) [reply src/code/z_scene.c:590]: Reworded it a little bit more as well.

File: src/code/z_scene.c (hunk 10 above)
Comment (hensldm) [reply src/code/z_scene.c:597]: Reworded it a little bit more as well.

File: src/code/z_demo.c [hunk 11]
```diff
@@ -660,11 +660,11 @@ void Cutscene_Command_ChooseCreditsScenes(PlayState* play, CutsceneContext* csCt
             case 2:
                 if (INV_CONTENT(ITEM_MASK_KAMARO) == ITEM_MASK_KAMARO) {
                     // Milk bar
-                    play->nextEntranceIndex = 0x2400;
+                    play->nextEntrance = ENTRANCE(MILK_BAR, 0);
                     gSaveContext.nextCutsceneIndex = 0xFFF0;
                 } else {
                     // SPOT00
-                    play->nextEntranceIndex = 0x1C00;
+                    play->nextEntrance = ENTRANCE(CUTSCENE, 0);
```
Comment (AngheloAlf) [inline src/code/z_demo.c:667]: I guess those comments can go away now

File: src/overlays/actors/ovl_En_Bji_01/z_en_bji_01.c [hunk 12]
```diff
@@ -342,7 +342,7 @@ void EnBji01_Init(Actor* thisx, PlayState* play) {
     SubS_FillCutscenesList(&this->actor, this->cutscenes, ARRAY_COUNT(this->cutscenes));
     this->moonsTear = (ObjMoonStone*)SubS_FindActor(play, NULL, ACTORCAT_PROP, ACTOR_OBJ_MOON_STONE);
 
-    switch (gSaveContext.save.entranceIndex) {
+    switch (gSaveContext.save.entrance) {
         case 0x4C00: /* Observatory from ECT */
         case 0x4C10: /* Observatory from Termina Field door */
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_En_Bji_01/z_en_bji_01.c:347]: Those cases should use the ENTRANCE macro too, right?

File: src/overlays/actors/ovl_En_Horse/z_en_horse.c [hunk 13]
```diff
@@ -1550,8 +1550,7 @@ void EnHorse_Stopping(EnHorse* this, PlayState* play) {
 
     if ((this->stateFlags & ENHORSE_STOPPING_NEIGH_SOUND) && (this->skin.skelAnime.curFrame > 29.0f)) {
         this->actor.speedXZ = 0.0f;
-        if ((Rand_ZeroOne() > 0.5f) &&
-            ((gSaveContext.save.entranceIndex != 0x6400) || !Cutscene_GetSceneSetupIndex(play))) {
+        if ((Rand_ZeroOne() > 0.5f) && ((gSaveContext.save.entrance != 0x6400) || !Cutscene_GetSceneSetupIndex(play))) {
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_En_Horse/z_en_horse.c:1553]: Missed entrance?
```suggestion
        if ((Rand_ZeroOne() > 0.5f) && ((gSaveContext.save.entrance != ENTRANCE(ROMANI_RANCH, 0)) || !Cutscene_GetSceneSetupIndex(play))) {
```

File: src/overlays/actors/ovl_En_Ma_Yto/z_en_ma_yto.c [hunk 14]
```diff
@@ -1092,10 +1092,10 @@ void EnMaYto_SetupPostMilkRunEnd(EnMaYto* this) {
 void EnMaYto_PostMilkRunEnd(EnMaYto* this, PlayState* play) {
     if (this->unk310 == 3) {
         // Termina Field
-        play->nextEntranceIndex = 0x54D0;
+        play->nextEntrance = ENTRANCE(TERMINA_FIELD, 13);
     } else {
         // Romani Ranch
-        play->nextEntranceIndex = 0x6480;
+        play->nextEntrance = ENTRANCE(ROMANI_RANCH, 8);
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_En_Ma_Yto/z_en_ma_yto.c:1098]: Those comments can go too

**Review by AngheloAlf — COMMENTED** [review]
the `gSaveContext.save.entranceIndex` -> `gSaveContext.save.entrance` and such should be added to namefixer

**Review by AngheloAlf — APPROVED** [review]

**Review by jpburnett — APPROVED** [review]

**Review by Kenix3 — APPROVED** [review]

**Pull request merged**